README.md
# bluse

> ⚗️ blend and fuse data with ease

[![Npm Version](https://img.shields.io/npm/v/bluse.svg)](https://www.npmjs.com/package/bluse)
[![Build Status](https://travis-ci.org/gabrielcsapo/bluse.svg?branch=master)](https://travis-ci.org/gabrielcsapo/bluse)
[![Coverage Status](https://lcov-server.herokuapp.com/badge/github%2Ecom/gabrielcsapo/monotime.svg)](https://lcov-server.herokuapp.com/coverage/github%2Ecom/gabrielcsapo/monotime)
[![Dependency Status](https://david-dm.org/gabrielcsapo/bluse.svg)](https://david-dm.org/gabrielcsapo/bluse)
[![devDependency Status](https://david-dm.org/gabrielcsapo/bluse/dev-status.svg)](https://david-dm.org/gabrielcsapo/bluse#info=devDependencies)
[![npm](https://img.shields.io/npm/dt/bluse.svg)](https://github.com/gabrielcsapo/bluse)
[![npm](https://img.shields.io/npm/dm/bluse.svg)](https://github.com/gabrielcsapo/bluse)

## Installation

```
npm install bluse --save
```

## Usage

> To use `bluse` as a standalone library without a bundler, use the bundle located at dist/`bluse`.js. Else if you are using webpack or any other bundler, simply require('`bluse`') will suffice!

```javascript
const bluse = require('bluse');

const data = [{
  builds: [{
    start: 1235,
    end: 12345,
    name: 'build1',
    steps: [{
      name: 'install',
      script: 'npm install',
      output: 'install successfully',
      dependencies: [
        'moment@1.0.2',
        'kraken-js@2.0.1',
        'tape@4.0.2'
      ],
      code: 0,
      start: 1,
      end: 34133
    }, {
      name: 'lint',
      script: 'npm run lint',
      output: 'failed lint!',
      code: 1,
      start: 2,
      end: 12
    }]
  }, {
    start: 1235,
    end: 12345,
    name: 'build2'
  }]
}, {
  builds: [{
    start: 1235,
    end: 12345,
    name: 'build3',
    steps: [{
      name: 'install',
      script: 'npm install',
      output: 'large outtttt',
      dependencies: [
        'moment@2.0.2',
        'kraken-js@2.0.1',
        'tape@10.0.2',
        'tap@11.0.2'
      ],
      code: 0,
      start: 1,
      end: 34123
    }, {
      name: 'lint',
      script: 'npm run lint',
      output: 'large outtttt',
      code: 0,
      start: 2,
      end: 1256
    }, {
      name: 'coverage',
      script: 'npm run coverage',
      output: 'RUNNING COVERAAAGGEE',
      code: 0,
      start: 2,
      end: 123456
    }]
  }, {
    start: 1235,
    end: 12345,
    name: 'build4'
  }]
}];

bluse(data, { unique: true }); // the default is false
```

Now that you have `blused` your data! It should look like the following:

```javascript
{
  'builds:start': [1235, 1235, 1235, 1235],
  'builds:end': [12345, 12345, 12345, 12345],
  'builds:name': ['build1', 'build2', 'build3', 'build4'],
  'builds:steps:name': ['install', 'lint', 'install', 'lint', 'coverage'],
  'builds:steps:script': ['npm install', 'npm run lint', 'npm install', 'npm run lint', 'npm run coverage'],
  'builds:steps:output': ['install successfully', 'failed lint!', 'large outtttt', 'large outtttt', 'RUNNING COVERAAAGGEE'],
  'builds:steps:dependencies': ['moment@1.0.2', 'kraken-js@2.0.1', 'tape@4.0.2', 'moment@2.0.2', 'kraken-js@2.0.1', 'tape@10.0.2', 'tap@11.0.2'],
  'builds:steps:code': [0, 1, 0, 0, 0],
  'builds:steps:start': [1, 2, 1, 2, 2],
  'builds:steps:end': [34133, 12, 34123, 1256, 123456]
}
```

## Contributions

`bluse` has been extensively tested with over 200,000 records and can hold up to the task! If you can break `bluse` please contribute back!


CHANGELOG.md
# 0.0.2 (09/25/2017)

- fixes issue with mixed array instead of assuming arrays are homogenous 

# 0.0.1 (09/25/2017)

- adds option to get unique values

# 0.0.0 (09/19/2017)

- basic usage works!


test/index.js
const test = require('tape');

const bluse = require('../index');

test('bluse', (t) => {
  t.plan(7);

  t.test('@object should flatten object', (t) => {
    const data = {
      name: "bluse",
      tags: ["nodejs", "utility", "npm"],
      description: "⚗️ blend and fuse data with ease",
      package: {
        directDependencies: 3,
        totalDependencies: 1093,
        resolveDependencies: [
          'moment@2.0.2',
          'tape@10.0.2',
          'tap@11.0.2'
        ]
      },
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build3'
      }, {
        start: 1235,
        end: 12345,
        name: 'build4'
      }]
    };

    t.deepEqual(bluse(data), {
      name: 'bluse',
      tags: ['nodejs', 'utility', 'npm'],
      description: '⚗️ blend and fuse data with ease',
      'package:directDependencies': 3,
      'package:totalDependencies': 1093,
      'package:resolveDependencies': ['moment@2.0.2', 'tape@10.0.2', 'tap@11.0.2'],
      'builds:start': [1235, 1235],
      'builds:end': [12345, 12345],
      'builds:name': ['build3', 'build4'],
      builds: []
    });
    t.end();
  })

  t.test('@array basic usage', (t) => {
    const data = [{
      name: "json-ex",
      tags: ['nodejs', 'utility', 'npm', 'json'],
      contributors: [{
        name: 'Gabriel J. Csapo'
      }],
      doesNotExistElsewhere: 'hello world',
      description: "Extends JSON to be able to serialize and deserialize more than just basic primitives.",
      package: {
        directDependencies: 45,
        totalDependencies: 803,
        resolveDependencies: [
          'moment@1.0.2',
          'tape@4.0.2'
        ]
      },
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build1'
      }, {
        start: 1235,
        end: 12345,
        name: 'build2'
      }]
    }, {
      name: "bluse",
      tags: ['nodejs', 'utility', 'npm'],
      contributors: [{
        name: 'Gabriel J. Csapo'
      }],
      description: "⚗️ blend and fuse data with ease",
      package: {
        directDependencies: 3,
        totalDependencies: 1093,
        resolveDependencies: [
          'moment@2.0.2',
          'tape@10.0.2',
          'tap@11.0.2'
        ]
      },
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build3'
      }, {
        start: 1235,
        end: 12345,
        name: 'build4'
      }]
    }];

    t.deepEqual(bluse(data), {
      name: ['json-ex', 'bluse'],
      tags: ['nodejs', 'utility', 'npm', 'json', 'nodejs', 'utility', 'npm'],
      'contributors:name': ['Gabriel J. Csapo', 'Gabriel J. Csapo'],
      contributors: [],
      doesNotExistElsewhere: ['hello world'],
      description: ['Extends JSON to be able to serialize and deserialize more than just basic primitives.', '⚗️ blend and fuse data with ease'],
      'package:directDependencies': [45, 3],
      'package:totalDependencies': [803, 1093],
      'package:resolveDependencies': ['moment@1.0.2', 'tape@4.0.2', 'moment@2.0.2', 'tape@10.0.2', 'tap@11.0.2'],
      'builds:start': [1235, 1235, 1235, 1235],
      'builds:end': [12345, 12345, 12345, 12345],
      'builds:name': ['build1', 'build2', 'build3', 'build4'],
      builds: []
    });
    t.end();
  });

  t.test('@array recursive nesting of arrays', (t) => {
    var data = [{
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build1',
        steps: [{
          name: 'install',
          script: 'npm install',
          output: 'install successfully',
          dependencies: [
            'moment@1.0.2',
            'kraken-js@2.0.1',
            'tape@4.0.2'
          ],
          code: 0,
          start: 1,
          end: 34133
        }, {
          name: 'lint',
          script: 'npm run lint',
          output: 'failed lint!',
          code: 1,
          start: 2,
          end: 12
        }]
      }, {
        start: 1235,
        end: 12345,
        name: 'build2'
      }]
    }, {
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build3',
        steps: [{
          name: 'install',
          script: 'npm install',
          output: 'large outtttt',
          dependencies: [
            'moment@2.0.2',
            'kraken-js@2.0.1',
            'tape@10.0.2',
            'tap@11.0.2'
          ],
          code: 0,
          start: 1,
          end: 34123
        }, {
          name: 'lint',
          script: 'npm run lint',
          output: 'large outtttt',
          code: 0,
          start: 2,
          end: 1256
        }, {
          name: 'coverage',
          script: 'npm run coverage',
          output: 'RUNNING COVERAAAGGEE',
          code: 0,
          start: 2,
          end: 123456
        }]
      }, {
        start: 1235,
        end: 12345,
        name: 'build4'
      }]
    }];

    t.deepEqual(bluse(data), {
      'builds:start': [1235, 1235, 1235, 1235],
      'builds:end': [12345, 12345, 12345, 12345],
      'builds:name': ['build1', 'build2', 'build3', 'build4'],
      'builds:steps:name': ['install', 'lint', 'install', 'lint', 'coverage'],
      'builds:steps:script': ['npm install', 'npm run lint', 'npm install', 'npm run lint', 'npm run coverage'],
      'builds:steps:output': ['install successfully', 'failed lint!', 'large outtttt', 'large outtttt', 'RUNNING COVERAAAGGEE'],
      'builds:steps:dependencies': ['moment@1.0.2', 'kraken-js@2.0.1', 'tape@4.0.2', 'moment@2.0.2', 'kraken-js@2.0.1', 'tape@10.0.2', 'tap@11.0.2'],
      'builds:steps:code': [0, 1, 0, 0, 0],
      'builds:steps:start': [1, 2, 1, 2, 2],
      'builds:steps:end': [34133, 12, 34123, 1256, 123456],
      'builds:steps': [],
      builds: []
    });
    t.end();
  });

  t.test('@array infinite nesting', (t) => {
    const data = [{
      account: {
        person: {
          name: {
            history: {
              date: 123456
            }
          },
          family: [{
            age: 50,
            name: 'Bob'
          }, {
            age: 28,
            name: 'Susan'
          }]
        }
      }
    }, {
      account: {
        person: {
          name: {
            history: {
              date: 654321
            }
          },
          family: [{
            age: 40,
            name: 'Diane'
          }, {
            age: 32,
            name: 'Alfred'
          }]
        }
      }
    }];

    t.deepEqual(bluse(data), {
      'account:person:name:history:date': [123456, 654321],
      'account:person:family:age': [50, 28, 40, 32],
      'account:person:family:name': ['Bob', 'Susan', 'Diane', 'Alfred'],
      'account:person:family': []
    });

    t.end();
  });

  t.test('@mixed should only return unique values', (t) => {
    const data = [{
      name: "json-ex",
      tags: ['nodejs', 'utility', 'npm', 'json', 'unique'],
      contributors: [{
        name: 'Gabriel J. Csapo'
      }],
      doesNotExistElsewhere: 'hello world',
      description: "Extends JSON to be able to serialize and deserialize more than just basic primitives.",
      package: {
        directDependencies: 45,
        totalDependencies: 803,
        resolveDependencies: [
          'moment@1.0.2',
          'tape@4.0.2'
        ]
      },
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build1'
      }, {
        start: 1235,
        end: 12345,
        name: 'build2'
      }]
    }, {
      name: "bluse",
      tags: ['nodejs', 'utility', 'npm'],
      contributors: [{
        name: 'Gabriel J. Csapo'
      }],
      description: "⚗️ blend and fuse data with ease",
      package: {
        directDependencies: 3,
        totalDependencies: 1093,
        resolveDependencies: [
          'moment@2.0.2',
          'tape@10.0.2',
          'tap@11.0.2'
        ]
      },
      builds: [{
        start: 1235,
        end: 12345,
        name: 'build3'
      }, {
        start: 1235,
        end: 12345,
        name: 'build4'
      }]
    }];

    t.deepEqual(bluse(data, {
      unique: true
    }), {
      name: ['json-ex', 'bluse'],
      tags: ['nodejs', 'utility', 'npm', 'json', 'unique'],
      'contributors:name': ['Gabriel J. Csapo'],
      contributors: [],
      doesNotExistElsewhere: ['hello world'],
      description: ['Extends JSON to be able to serialize and deserialize more than just basic primitives.', '⚗️ blend and fuse data with ease'],
      'package:directDependencies': [45, 3],
      'package:totalDependencies': [803, 1093],
      'package:resolveDependencies': ['moment@1.0.2', 'tape@4.0.2', 'moment@2.0.2', 'tape@10.0.2', 'tap@11.0.2'],
      'builds:start': [1235],
      'builds:end': [12345],
      'builds:name': ['build1', 'build2', 'build3', 'build4'],
      builds: []
    });
    t.end();
  });

  t.test('@mixed should be able to handled mixed arrays', (t) => {
    var data = {
      foo: ['echo $FOO'],
      install: ['npm --version', 'node --version', {
        npm: []
      }],
      lint: ['npm run lint'],
      coverage: ['npm run coverage'],
      test: ['npm test'],
      docs: ['npm run generate-docs']
    };

    t.deepEqual(bluse(data), {
      foo: ['echo $FOO'],
      'install:npm': [],
      install: ['npm --version', 'node --version'],
      lint: ['npm run lint'],
      coverage: ['npm run coverage'],
      test: ['npm test'],
      docs: ['npm run generate-docs']
    });
    t.end();
  });

  t.test('@mixed should be able to handled mixed arrays nested inside mixed arrays', (t) => {
    var data = {
      foo: ['echo $FOO'],
      install: ['npm --version', 'node --version', {
        npm: ['npm install', {
          list: ['ls -lh']
        }]
      }],
      lint: ['npm run lint'],
      coverage: ['npm run coverage'],
      test: ['npm test'],
      docs: ['npm run generate-docs']
    };

    t.deepEqual(bluse(data), {
      foo: ['echo $FOO'],
      'install:npm:list': ['ls -lh'],
      'install:npm': ['npm install'],
      install: ['npm --version', 'node --version'],
      lint: ['npm run lint'],
      coverage: ['npm run coverage'],
      test: ['npm test'],
      docs: ['npm run generate-docs']
    });
    t.end();
  });

});


index.js
function parser(data, pk) {
  let struct = {};

  Object.keys(data).forEach((k) => {
    if(Array.isArray(data[k]) == true) {
      // deal with mixed array values by separating them
      let objects = [];
      let flat = [];
      data[k].forEach((o) => {
        if(typeof o === 'object') {
          objects.push(o);
        } else {
          flat.push(o);
        }
      });

      if(objects.length > 0) {
        // if this array contains objects, parse them accordingly
        var na = objects.map((sd) => parser(sd, pk ? `${pk}:${k}` : k));
        na.forEach((n, i) => {
          Object.keys(n).forEach((a) => {
            // the key is since we have already run the parser on the data
            // we are simply putting it back on the top level object
            if(struct[a]) {
              struct[a] = struct[a].concat(na[i][a]);
            } else {
              if(Array.isArray(na[i][a])) {
                struct[a] = na[i][a];
              } else {
                struct[a] = [na[i][a]];
              }
            }
          });
        });
      }
      // if this is a flat array, just put the data back
      struct[pk ? `${pk}:${k}` : k] = flat;

    } else if(typeof data[k] == 'object') {
      Object.assign(struct, parser(data[k], pk ? `${pk}:${k}` : k));
    } else {
      // this is a flat attribute
      struct[pk ? `${pk}:${k}` : k] = data[k];
    }
  });

  return struct;
}

function fUnique(value, index, self) {
    return self.indexOf(value) === index;
}

module.exports = (data, options={}) => {
  const { unique=false } = options;

  if(Array.isArray(data)) {
    let agr = {};
    const ret = data.map((d) => parser(d));

    ret.forEach((r) => {
      Object.keys(r).forEach((k) => {
        if(agr[k]) {
          if(unique) {
            agr[k] = agr[k].concat(r[k]).filter(fUnique);
          } else {
            agr[k] = agr[k].concat(r[k]);
          }
        } else {
          if(Array.isArray(r[k])) {
            if(unique) {
              agr[k] = r[k].filter(fUnique);
            } else {
              agr[k] = r[k];
            }
          } else {
            if(unique) {
              agr[k] = [r[k]].filter(fUnique);
            } else {
              agr[k] = [r[k]]
            }
          }
        }
      })
    });
    return agr;
  }
  return parser(data);
}


tryitout.js
module.exports = {
    title: "bluse",
    description: "Generating data is hard enough, combining it should be easy ⚗️",
    nav: {
      Source: "https://github.com/gabrielcsapo/bluse"
    },
    options: {
      width: '90%'
    },
    body: [{
      type: "html",
      value: `
        <div style="text-align:center;">
          <blockquote style="display:inline-block;">
            <a href="https://github.com/gabrielcsapo/bluse"><code>bluse</code></a> ⚗️ blend and fuse data with ease
          </blockquote>
        </div>
      `
    }, {
      type: "code",
      title: "Bluse a simple object 😉",
      value: `
        const bluse = require('bluse');

        const data = {
          name: "bluse",
          tags: ["nodejs", "utility", "npm"],
          description: "⚗️ blend and fuse data",
          package: {
            directDependencies: 3,
            totalDependencies: 1093,
            resolveDependencies: [
              'moment@2.0.2',
              'tape@10.0.2',
              'tap@11.0.2'
            ]
          },
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build3'
          }, {
            start: 1235,
            end: 12345,
            name: 'build4'
          }]
        };

        console.log(JSON.stringify(bluse(data), null, 4));
      `
    }, {
      type: 'code',
      title: 'Bluse an array of complex objects 🔬',
      value: `
        const bluse = require('bluse');

        const data = [{
          name: "json-ex",
          tags: ['nodejs', 'utility', 'npm', 'json'],
          contributors: [{
            name: 'Gabriel J. Csapo'
          }],
          doesNotExistElsewhere: 'hello world',
          description: "Extends JSON to be able to serialize and deserialize more than just basic primitives.",
          package: {
            directDependencies: 45,
            totalDependencies: 803,
            resolveDependencies: [
              'moment@1.0.2',
              'tape@4.0.2'
            ]
          },
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build1'
          }, {
            start: 1235,
            end: 12345,
            name: 'build2'
          }]
        }, {
          name: "bluse",
          tags: ['nodejs', 'utility', 'npm'],
          contributors: [{
            name: 'Gabriel J. Csapo'
          }],
          description: "⚗️ blend and fuse data",
          package: {
            directDependencies: 3,
            totalDependencies: 1093,
            resolveDependencies: [
              'moment@2.0.2',
              'tape@10.0.2',
              'tap@11.0.2'
            ]
          },
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build3'
          }, {
            start: 1235,
            end: 12345,
            name: 'build4'
          }]
        }];

        console.log(JSON.stringify(bluse(data), null, 4));
      `
    }, {
      type: 'code',
      title: 'Bluse an array of EXTREMELY complex objects 🔬',
      value: `
        const bluse = require('bluse');

        const data = [{
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build1',
            steps: [{
              name: 'install',
              script: 'npm install',
              output: 'install successfully',
              dependencies: [
                'moment@1.0.2',
                'kraken-js@2.0.1',
                'tape@4.0.2'
              ],
              code: 0,
              start: 1,
              end: 34133
            }, {
              name: 'lint',
              script: 'npm run lint',
              output: 'failed lint!',
              code: 1,
              start: 2,
              end: 12
            }]
          }, {
            start: 1235,
            end: 12345,
            name: 'build2'
          }]
        }, {
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build3',
            steps: [{
              name: 'install',
              script: 'npm install',
              output: 'large outtttt',
              dependencies: [
                'moment@2.0.2',
                'kraken-js@2.0.1',
                'tape@10.0.2',
                'tap@11.0.2'
              ],
              code: 0,
              start: 1,
              end: 34123
            }, {
              name: 'lint',
              script: 'npm run lint',
              output: 'large outtttt',
              code: 0,
              start: 2,
              end: 1256
            }, {
              name: 'coverage',
              script: 'npm run coverage',
              output: 'RUNNING COVERAAAGGEE',
              code: 0,
              start: 2,
              end: 123456
            }]
          }, {
            start: 1235,
            end: 12345,
            name: 'build4'
          }]
        }];

        console.log(JSON.stringify(bluse(data), null, 4));
      `
    }, {
      type: 'code',
      title: 'Bluse an array of complex objects and get back unique values 🔬',
      value: `
        const bluse = require('bluse');

        const data = [{
          name: "json-ex",
          tags: ['nodejs', 'utility', 'npm', 'json'],
          contributors: [{
            name: 'Gabriel J. Csapo'
          }],
          doesNotExistElsewhere: 'hello world',
          description: "Extends JSON to be able to serialize and deserialize more than just basic primitives.",
          package: {
            directDependencies: 45,
            totalDependencies: 803,
            resolveDependencies: [
              'moment@1.0.2',
              'tape@4.0.2'
            ]
          },
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build1'
          }, {
            start: 1235,
            end: 12345,
            name: 'build2'
          }]
        }, {
          name: "bluse",
          tags: ['nodejs', 'utility', 'npm'],
          contributors: [{
            name: 'Gabriel J. Csapo'
          }],
          description: "⚗️ blend and fuse data",
          package: {
            directDependencies: 3,
            totalDependencies: 1093,
            resolveDependencies: [
              'moment@2.0.2',
              'tape@10.0.2',
              'tap@11.0.2'
            ]
          },
          builds: [{
            start: 1235,
            end: 12345,
            name: 'build3'
          }, {
            start: 1235,
            end: 12345,
            name: 'build4'
          }]
        }];

        console.log(JSON.stringify(bluse(data, { unique: true }), null, 4));
      `
    }, {
      type: 'code',
      title: 'Bluse an object with nested mixed array types 🔬',
      value: `
        const bluse = require('bluse');

        const data = {
          foo: ['echo $FOO'],
          bar: ['bar', {
            barfoo: []
          }],
          install: ['npm --version', 'node --version', {
            npm: ['npm install', {
              list: ['ls -lh']
            }]
          }],
          lint: ['npm run lint'],
          coverage: ['npm run coverage'],
          test: ['npm test'],
          docs: ['npm run generate-docs']
        };

        console.log(JSON.stringify(bluse(data), null, 4));
      `
    }],
    output: "./docs",
    externals: [
      "./dist/bluse.min.js"
    ]
}
